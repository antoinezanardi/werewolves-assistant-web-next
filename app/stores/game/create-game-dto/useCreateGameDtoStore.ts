import { defineStore } from "pinia";
import { ref } from "vue";
import { DEFAULT_GAME_OPTIONS } from "~/composables/api/game/constants/game-options/game-options.constants";
import type { CreateGameAdditionalCardDto } from "~/composables/api/game/dto/create-game/create-game-additional-card/create-game-additional-card.dto";

import { CreateGamePlayerDto } from "~/composables/api/game/dto/create-game/create-game-player/create-game-player.dto";
import { CreateGameDto } from "~/composables/api/game/dto/create-game/create-game.dto";
import type { GameAdditionalCardRecipientRoleName } from "~/composables/api/game/types/game-additional-card/game-additional-card.types";
import { GameOptions } from "~/composables/api/game/types/game-options/game-options.class";
import { ADDITIONAL_CARDS_DEPENDANT_ROLES } from "~/composables/api/role/constants/role.constants";
import type { RoleName } from "~/composables/api/role/types/role.types";
import { StoreIds } from "~/stores/enums/store.enum";
import { useRolesStore } from "~/stores/role/useRolesStore";
import { useLocalStorage } from "@vueuse/core";
import { LocalStorageKeys } from "~/utils/enums/local-storage.enums";

const useCreateGameDtoStore = defineStore(StoreIds.CREATE_GAME_DTO, () => {
  const rolesStore = useRolesStore();
  const { getRoleWithNameInRoles } = rolesStore;

  const createGameOptionsDtoFromLocalStorage = useLocalStorage(LocalStorageKeys.GAME_OPTIONS, DEFAULT_GAME_OPTIONS, { mergeDefaults: true });

  const defaultCreateGameDto = CreateGameDto.create({
    players: [],
    options: DEFAULT_GAME_OPTIONS,
  });

  const createGameDto = ref<CreateGameDto>(CreateGameDto.create({
    ...defaultCreateGameDto,
    options: createGameOptionsDtoFromLocalStorage.value,
  }));

  const createGameOptionsDto = computed<GameOptions>(() => createGameDto.value.options);

  const doesCreateGameDtoContainPositionDependantRoles = computed<boolean>(() => {
    const positionDependantRoles: RoleName[] = ["rusty-sword-knight", "bear-tamer", "fox"];
    const playersWithPositionDependantRoles = getPlayersWithAnyRoleNameInCreateGameDto(positionDependantRoles);

    return playersWithPositionDependantRoles.length > 0;
  });

  const doesCreateGameDtoContainAdditionalCardsDependantRoles = computed<boolean>(() => {
    const playersWithAdditionalCardsDependantRoles = getPlayersWithAnyRoleNameInCreateGameDto([...ADDITIONAL_CARDS_DEPENDANT_ROLES]);

    return playersWithAdditionalCardsDependantRoles.length > 0;
  });

  function setCreateGameDto(createGameDtoValue: CreateGameDto): void {
    createGameDto.value = CreateGameDto.create(createGameDtoValue);
    saveCreateGameOptionsDtoToLocalStorage();
  }

  function resetCreateGameDto(doesRetrieveLocalStorageValues = true): void {
    if (!doesRetrieveLocalStorageValues) {
      createGameDto.value = CreateGameDto.create(defaultCreateGameDto);
      saveCreateGameOptionsDtoToLocalStorage();

      return;
    }
    createGameDto.value = CreateGameDto.create({
      ...defaultCreateGameDto,
      options: createGameOptionsDtoFromLocalStorage.value,
    });
  }

  function resetCreateGameOptionsDto(): void {
    createGameDto.value.options = GameOptions.create(DEFAULT_GAME_OPTIONS);
    saveCreateGameOptionsDtoToLocalStorage();
  }

  function saveCreateGameOptionsDtoToLocalStorage(): void {
    createGameOptionsDtoFromLocalStorage.value = createGameDto.value.options;
  }

  function removeObsoleteAdditionalCardsFromCreateGameDto(): void {
    if (!createGameDto.value.additionalCards) {
      return;
    }
    const rolesInGame = new Set(createGameDto.value.players.map(player => player.role.name));
    createGameDto.value.additionalCards = createGameDto.value.additionalCards.filter(({ recipient }) => rolesInGame.has(recipient));
  }

  function addPlayerToCreateGameDto(player: CreateGamePlayerDto): void {
    createGameDto.value.players.push(CreateGamePlayerDto.create(player));
  }

  function updatePlayerInCreateGameDto(player: CreateGamePlayerDto): void {
    const playerIndex = createGameDto.value.players.findIndex(({ name }) => name === player.name);
    if (playerIndex !== -1) {
      createGameDto.value.players.splice(playerIndex, 1, CreateGamePlayerDto.create(player));
    }
  }

  function setPlayersToCreateGameDto(players: CreateGamePlayerDto[]): void {
    createGameDto.value.players = players.map(player => CreateGamePlayerDto.create(player));
  }

  function removePlayerFromCreateGameDto(playerName: string): void {
    const playerIndex = createGameDto.value.players.findIndex(player => player.name === playerName);
    if (playerIndex !== -1) {
      createGameDto.value.players = createGameDto.value.players.toSpliced(playerIndex, 1);
    }
  }

  function isRoleInCreateGameDto(roleName: RoleName): boolean {
    return createGameDto.value.players.some(player => player.role.name === roleName);
  }

  function getPlayersWithRoleNameInCreateGameDto(roleName: RoleName): CreateGamePlayerDto[] {
    return createGameDto.value.players.filter(player => player.role.name === roleName);
  }

  function getPlayersWithAnyRoleNameInCreateGameDto(roleNames: RoleName[]): CreateGamePlayerDto[] {
    return createGameDto.value.players.filter(player => player.role.name && roleNames.includes(player.role.name));
  }

  function isRoleMinReachedInCreateGameDto(roleName: RoleName): boolean {
    const playersWithRoleName = getPlayersWithRoleNameInCreateGameDto(roleName);
    const role = getRoleWithNameInRoles(roleName);

    return !!role && (role.minInGame === undefined || playersWithRoleName.length >= role.minInGame);
  }

  function isRoleMaxReachedInCreateGameDto(roleName: RoleName): boolean {
    const playersWithRoleName = getPlayersWithRoleNameInCreateGameDto(roleName);
    const role = getRoleWithNameInRoles(roleName);

    return !!role && playersWithRoleName.length >= role.maxInGame;
  }

  function getRoleLeftCountToReachMinInCreateGameDto(roleName: RoleName): number {
    const playersWithRoleName = getPlayersWithRoleNameInCreateGameDto(roleName);
    const role = getRoleWithNameInRoles(roleName);
    if (role?.minInGame === undefined) {
      return 0;
    }
    const leftCount = role.minInGame - playersWithRoleName.length;

    return leftCount <= -1 ? 0 : leftCount;
  }

  function setAdditionalCardsForRecipientInCreateGameDto(additionalCards: CreateGameAdditionalCardDto[], recipient: GameAdditionalCardRecipientRoleName): void {
    if (!createGameDto.value.additionalCards) {
      createGameDto.value.additionalCards = additionalCards;
    }

    createGameDto.value.additionalCards = createGameDto.value.additionalCards.filter(card => card.recipient !== recipient);
    createGameDto.value.additionalCards.push(...additionalCards);
  }

  function getAdditionalCardsForRecipientInCreateGameDto(recipient: GameAdditionalCardRecipientRoleName): CreateGameAdditionalCardDto[] {
    return createGameDto.value.additionalCards?.filter(card => card.recipient === recipient) ?? [];
  }

  function getAdditionalCardsWithRoleNameInCreateGameDto(roleName: RoleName): CreateGameAdditionalCardDto[] {
    return createGameDto.value.additionalCards?.filter(card => card.roleName === roleName) ?? [];
  }
  return {
    createGameOptionsDtoFromLocalStorage,
    createGameDto,
    createGameOptionsDto,
    doesCreateGameDtoContainPositionDependantRoles,
    doesCreateGameDtoContainAdditionalCardsDependantRoles,
    setCreateGameDto,
    resetCreateGameDto,
    resetCreateGameOptionsDto,
    saveCreateGameOptionsDtoToLocalStorage,
    removeObsoleteAdditionalCardsFromCreateGameDto,
    addPlayerToCreateGameDto,
    updatePlayerInCreateGameDto,
    setPlayersToCreateGameDto,
    removePlayerFromCreateGameDto,
    isRoleInCreateGameDto,
    getPlayersWithRoleNameInCreateGameDto,
    getPlayersWithAnyRoleNameInCreateGameDto,
    isRoleMinReachedInCreateGameDto,
    isRoleMaxReachedInCreateGameDto,
    getRoleLeftCountToReachMinInCreateGameDto,
    setAdditionalCardsForRecipientInCreateGameDto,
    getAdditionalCardsForRecipientInCreateGameDto,
    getAdditionalCardsWithRoleNameInCreateGameDto,
  };
});

export { useCreateGameDtoStore };