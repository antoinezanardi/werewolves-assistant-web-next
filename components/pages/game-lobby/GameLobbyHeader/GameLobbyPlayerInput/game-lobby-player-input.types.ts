type GameLobbyPlayerInputProps = {
  modelValue: string;
};

export type { GameLobbyPlayerInputProps };