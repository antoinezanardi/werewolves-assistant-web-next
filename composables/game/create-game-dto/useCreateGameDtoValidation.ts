import type { ComputedRef, Ref } from "vue";

import { MIN_PLAYERS_IN_GAME } from "~/composables/api/game/constants/game.constants";
import type { CreateGameDto } from "~/composables/api/game/dto/create-game/create-game.dto";

type UseCreateGameDtoValidation = {
  isMinimumPlayersReached: ComputedRef<boolean>;
  areAllRolesSet: ComputedRef<boolean>;
  canCreateGame: ComputedRef<boolean>;
};

function useCreateGameDtoValidation(createGameDto: Ref<CreateGameDto>): UseCreateGameDtoValidation {
  const isMinimumPlayersReached = computed<boolean>(() => createGameDto.value.players.length >= MIN_PLAYERS_IN_GAME);

  const areAllRolesSet = computed<boolean>(() => createGameDto.value.players.every(player => player.role.name !== undefined));

  const canCreateGame = computed<boolean>(() => isMinimumPlayersReached.value && areAllRolesSet.value);

  return {
    isMinimumPlayersReached,
    areAllRolesSet,
    canCreateGame,
  };
}

export { useCreateGameDtoValidation };