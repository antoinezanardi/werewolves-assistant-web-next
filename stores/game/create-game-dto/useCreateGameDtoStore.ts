import { defineStore } from "pinia";
import { ref } from "vue";

import { CreateGamePlayerDto } from "~/composables/api/game/dto/create-game/create-game-player/create-game-player.dto";
import { CreateGameDto } from "~/composables/api/game/dto/create-game/create-game.dto";
import { StoreIds } from "~/stores/enums/store.enum";

const useCreateGameDtoStore = defineStore(StoreIds.CREATE_GAME_DTO, () => {
  const createGameDto = ref<CreateGameDto>(CreateGameDto.create({ players: [] }));

  function setCreateGameDto(createGameDtoValue: CreateGameDto): void {
    createGameDto.value = CreateGameDto.create(createGameDtoValue);
  }

  function resetCreateGameDto(): void {
    createGameDto.value = CreateGameDto.create({ players: [] });
  }

  function addPlayerToCreateGameDto(player: CreateGamePlayerDto): void {
    createGameDto.value.players.push(CreateGamePlayerDto.create(player));
  }

  function updatePlayerInCreateGameDto(player: CreateGamePlayerDto): void {
    const playerIndex = createGameDto.value.players.findIndex(({ name }) => name === player.name);
    if (playerIndex !== -1) {
      createGameDto.value.players.splice(playerIndex, 1, CreateGamePlayerDto.create(player));
    }
  }

  function setPlayersToCreateGameDto(players: CreateGamePlayerDto[]): void {
    createGameDto.value.players = players.map(player => CreateGamePlayerDto.create(player));
  }

  function removePlayerFromCreateGameDto(playerName: string): void {
    const playerIndex = createGameDto.value.players.findIndex(player => player.name === playerName);
    if (playerIndex !== -1) {
      createGameDto.value.players = createGameDto.value.players.toSpliced(playerIndex, 1);
    }
  }
  return {
    createGameDto,
    setCreateGameDto,
    resetCreateGameDto,
    addPlayerToCreateGameDto,
    updatePlayerInCreateGameDto,
    setPlayersToCreateGameDto,
    removePlayerFromCreateGameDto,
  };
});

export { useCreateGameDtoStore };