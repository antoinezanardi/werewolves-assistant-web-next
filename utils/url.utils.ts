function removeTrailingSlashes(url: string): string {
  return url.replace(/\/+$/u, "");
}

export { removeTrailingSlashes };