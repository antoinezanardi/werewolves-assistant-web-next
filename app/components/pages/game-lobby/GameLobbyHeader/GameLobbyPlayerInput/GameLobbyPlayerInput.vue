<template>
  <div id="game-lobby-player-input">
    <PrimeVueInputGroup
      id="game-lobby-player-input-group"
      class="mt-3"
    >
      <PrimeVueFloatLabel>
        <PrimeVueInputText
          id="player-name-input"
          ref="playerNameInput"
          v-model="inputValue"
          aria-labelledby="player-name-input-help"
          :class="{ 'p-invalid': doesPlayerNameExistInGame }"
          :disabled="isInputDisabled"
          :maxlength="MAX_PLAYER_NAME_LENGTH"
          size="large"
          type="text"
        />

        <label for="player-name-input">
          {{ $t("components.GameLobbyPlayerInput.playerName") }}
        </label>
      </PrimeVueFloatLabel>

      <PrimeVueButton
        id="add-player-button"
        :disabled="isAddButtonDisabled"
        :label="$t('shared.actions.add')"
        severity="primary"
        type="submit"
      >
        <template #icon>
          <FontAwesomeIcon icon="plus"/>
        </template>
      </PrimeVueButton>
    </PrimeVueInputGroup>

    <Transition
      mode="out-in"
      name="fade"
    >
      <small
        id="player-name-input-help"
        :key="playerNameInputHelpText"
      >
        {{ playerNameInputHelpText }}
      </small>
    </Transition>
  </div>
</template>

<script lang="ts" setup>
import { FontAwesomeIcon } from "@fortawesome/vue-fontawesome";
import { storeToRefs } from "pinia";
import { computed } from "vue";
import type { ComponentPublicInstance } from "vue";

import { MAX_PLAYERS_IN_GAME } from "~/composables/api/game/constants/game.constants";
import { MAX_PLAYER_NAME_LENGTH } from "~/composables/api/game/constants/player/player.constants";
import { useDevice } from "~/composables/misc/useDevice";
import { useCreateGameDtoStore } from "~/stores/game/create-game-dto/useCreateGameDtoStore";

const { t } = useI18n();

const createGameDtoStore = useCreateGameDtoStore();

const { isOnTouchDevice } = useDevice();

const { createGameDto } = storeToRefs(createGameDtoStore);

const inputValue = defineModel<string>({ required: true });

const playerNameInput = ref<ComponentPublicInstance | null>(null);

const doesPlayerNameExistInGame = computed<boolean>(() => createGameDto.value.players.some(({ name }) => name === inputValue.value.trim()));

const hasPlayerNameReachedMaxLength = computed<boolean>(() => inputValue.value.trim().length >= MAX_PLAYER_NAME_LENGTH);

const doesGameHaveMaxPlayers = computed<boolean>(() => createGameDto.value.players.length >= MAX_PLAYERS_IN_GAME);

const isInputDisabled = computed<boolean>(() => doesGameHaveMaxPlayers.value);

const isAddButtonDisabled = computed<boolean>(() => doesPlayerNameExistInGame.value || isInputDisabled.value);

const playerNameInputHelpText = computed<string>(() => {
  if (doesGameHaveMaxPlayers.value) {
    return t("components.GameLobbyPlayerInput.maxPlayersReached");
  }
  if (doesPlayerNameExistInGame.value) {
    return t("components.GameLobbyPlayerInput.playerNameIsAlreadyTaken");
  }
  if (hasPlayerNameReachedMaxLength.value) {
    return t("components.GameLobbyPlayerInput.playerNameMaxLengthReached");
  }
  return t("components.GameLobbyPlayerInput.pleaseEnterPlayerName");
});

function focusOnPlayerNameInput(): void {
  if (!playerNameInput.value) {
    throw createError("Player name input is not defined");
  }
  if (!isOnTouchDevice.value) {
    (playerNameInput.value.$el as HTMLElement).focus();
  }
}

onMounted(focusOnPlayerNameInput);

defineExpose({ isAddButtonDisabled });
</script>