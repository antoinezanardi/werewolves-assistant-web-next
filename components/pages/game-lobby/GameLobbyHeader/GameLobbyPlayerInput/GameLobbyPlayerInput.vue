<template>
  <VuePrimeInputGroup
    id="game-lobby-player-input-group"
    class="mt-3"
  >
    <span class="p-float-label">
      <VuePrimeInputText
        id="player-name-input"
        v-model="inputValue"
        :disabled="isInputDisabled"
        size="large"
        type="text"
      />

      <label for="player-name-input">
        {{ $t("components.GameLobbyPlayerInput.playerName") }}
      </label>
    </span>

    <VuePrimeButton
      id="add-player-button"
      :disabled="isAddButtonDisabled"
      role="submit"
      severity="primary"
    >
      <i class="fa fa-plus me-2"/>

      <span>
        {{ $t("shared.actions.add") }}
      </span>
    </VuePrimeButton>
  </VuePrimeInputGroup>
</template>

<script lang="ts" setup>
import { storeToRefs } from "pinia";
import { computed, defineModel } from "vue";

import { MAX_PLAYERS_IN_GAME } from "~/composables/api/game/constants/game.constants";
import { useCreateGameDtoStore } from "~/stores/game/create-game-dto/useCreateGameDtoStore";

const createGameDtoStore = useCreateGameDtoStore();

const { createGameDto } = storeToRefs(createGameDtoStore);

const inputValue = defineModel<string>({ required: true });

const isInputDisabled = computed<boolean>(() => createGameDto.value.players.length >= MAX_PLAYERS_IN_GAME);

const isAddButtonDisabled = computed<boolean>(() => createGameDto.value.players.some(({ name }) => name === inputValue.value.trim()) || isInputDisabled.value);
</script>