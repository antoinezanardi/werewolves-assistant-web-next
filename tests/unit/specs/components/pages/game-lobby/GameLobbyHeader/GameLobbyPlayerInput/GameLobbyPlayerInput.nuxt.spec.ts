import type { mount } from "@vue/test-utils";
import type { ComponentMountingOptions } from "@vue/test-utils/dist/mount";
import type InputText from "primevue/inputtext";

import type { GameLobbyPlayerInputProps } from "~/components/pages/game-lobby/GameLobbyHeader/GameLobbyPlayerInput/game-lobby-player-input.types";
import GameLobbyPlayerInput from "~/components/pages/game-lobby/GameLobbyHeader/GameLobbyPlayerInput/GameLobbyPlayerInput.vue";
import { useCreateGameDtoStore } from "~/stores/game/create-game-dto/useCreateGameDtoStore";
import { createFakeCreateGamePlayerDto } from "~/tests/unit/utils/factories/composables/api/game/dto/create-game/create-game-player/create-game-player.dto.factory";
import { mountSuspendedComponent } from "~/tests/unit/utils/mount.utils";

describe("Game Lobby Player Input Component", () => {
  let wrapper: ReturnType<typeof mount<typeof GameLobbyPlayerInput>>;
  const defaultProps: GameLobbyPlayerInputProps = { modelValue: "" };
  const defaultMountingOptions: ComponentMountingOptions<typeof GameLobbyPlayerInput> = {
    props: defaultProps,
    global: { stubs: { InputGroup: false } },
  };

  beforeEach(async() => {
    wrapper = await mountSuspendedComponent(GameLobbyPlayerInput, defaultMountingOptions);
  });

  it("should render component and match snapshot when mounted.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("Text input", () => {
    it("should be prefilled when v-model value is not empty.", async() => {
      wrapper = await mountSuspendedComponent(GameLobbyPlayerInput, {
        ...defaultMountingOptions,
        props: { modelValue: "Player 1" },
      });
      const input = wrapper.findComponent<typeof InputText>("#player-name-input");

      expect(input.attributes("modelvalue")).toBe("Player 1");
    });

    it("should be disabled when create game dto has reached max players.", async() => {
      const createGameDtoStore = useCreateGameDtoStore();
      createGameDtoStore.createGameDto.players = Array.from({ length: 40 }, createFakeCreateGamePlayerDto);
      const input = wrapper.findComponent<typeof InputText>("#player-name-input");
      await nextTick();

      expect(input.attributes("disabled")).toBe("true");
    });

    it("should not be disabled when create game dto has not reached max players.", async() => {
      const createGameDtoStore = useCreateGameDtoStore();
      createGameDtoStore.createGameDto.players = Array.from({ length: 39 }, createFakeCreateGamePlayerDto);
      const input = wrapper.findComponent<typeof InputText>("#player-name-input");
      await nextTick();

      expect(input.attributes("disabled")).toBe("false");
    });

    it("should translate input label when rendered.", () => {
      const label = wrapper.find<HTMLLabelElement>("[for='player-name-input']");

      expect(label.text()).toBe("Player name");
    });
  });

  describe("Add button", () => {
    it("should be disabled when create game dto has reached max players.", async() => {
      const createGameDtoStore = useCreateGameDtoStore();
      createGameDtoStore.createGameDto.players = Array.from({ length: 40 }, createFakeCreateGamePlayerDto);
      const button = wrapper.find<HTMLButtonElement>("#add-player-button");
      await nextTick();

      expect(button.attributes("disabled")).toBe("true");
    });

    it("should be disabled when one player name is same as input value.", async() => {
      const createGameDtoStore = useCreateGameDtoStore();
      createGameDtoStore.createGameDto.players = [
        createFakeCreateGamePlayerDto({ name: "Player 1" }),
        createFakeCreateGamePlayerDto({ name: "Player 2" }),
        createFakeCreateGamePlayerDto({ name: "Player 3" }),
      ];
      const input = wrapper.findComponent<typeof InputText>("#player-name-input");
      const button = wrapper.find<HTMLButtonElement>("#add-player-button");
      await input.setValue("Player 1");

      expect(button.attributes("disabled")).toBe("true");
    });

    it("should be disabled when one player name is same as input value even with some spaces around it.", async() => {
      const createGameDtoStore = useCreateGameDtoStore();
      createGameDtoStore.createGameDto.players = [
        createFakeCreateGamePlayerDto({ name: "Player 1" }),
        createFakeCreateGamePlayerDto({ name: "Player 2" }),
        createFakeCreateGamePlayerDto({ name: "Player 3" }),
      ];
      const input = wrapper.findComponent<typeof InputText>("#player-name-input");
      const button = wrapper.find<HTMLButtonElement>("#add-player-button");
      await input.setValue("   Player 1   ");

      expect(button.attributes("disabled")).toBe("true");
    });

    it("should be enabled when input value is not empty and create game dto has not reached max players.", async() => {
      const createGameDtoStore = useCreateGameDtoStore();
      createGameDtoStore.createGameDto.players = Array.from({ length: 39 }, createFakeCreateGamePlayerDto);
      const input = wrapper.findComponent<typeof InputText>("#player-name-input");
      const button = wrapper.find<HTMLButtonElement>("#add-player-button");
      await input.setValue("Player 2");

      expect(button.attributes("disabled")).toBe("false");
    });

    it("should translate button label when rendered.", async() => {
      wrapper = await mountSuspendedComponent(GameLobbyPlayerInput, {
        ...defaultMountingOptions,
        global: {
          stubs: {
            InputGroup: false,
            Button: false,
          },
        },
      });
      const button = wrapper.find<HTMLButtonElement>("#add-player-button");

      expect(button.text()).toBe("Add");
    });
  });
});