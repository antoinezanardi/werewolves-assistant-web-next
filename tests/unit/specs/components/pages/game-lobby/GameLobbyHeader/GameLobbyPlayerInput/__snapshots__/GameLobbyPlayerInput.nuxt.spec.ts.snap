// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game Lobby Player Input Component > should render component and match snapshot when mounted. 1`] = `
"<div class="p-inputgroup mt-3" data-pc-name="inputgroup" data-pc-section="root" id="game-lobby-player-input-group"><span class="p-float-label"><input-text-stub id="player-name-input" modelvalue="" disabled="false" size="large" type="text"></input-text-stub><label for="player-name-input">Player name</label></span>
  <button-stub id="add-player-button" disabled="false" role="submit" severity="primary"></button-stub>
</div>"
`;
